File: tree/leetcode_Sum_Root_to_Leaf_Numbers.java
/*Sum_Root_to_Leaf_Numbers

Given a binary tree containing digits from 0-9 only, each root-to-leaf path
could represent a number.
An example is the root-to-leaf path 1->2->3 which represents the number 123.
Find the total sum of all root-to-leaf numbers.
For example,

        1
       / \
      2   3
The root-to-leaf path 1->2 represents the number 12.
The root-to-leaf path 1->3 represents the number 13.

Return the sum = 12 + 13 = 25.
*/

import java.util.ArrayList;


public class leetcode_Sum_Root_to_Leaf_Numbers {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(5);
        root.right = new TreeNode(1);
        root.right.right = new TreeNode(6);
        System.out.println(sumNumbers(root));
    }

    public static int sumNumbers(TreeNode root) {
        // First impression. DFS + tracking cur sum.
        return preorder(root, 0);
    }

    public int preorder(TreeNode root, int sum) {
        if (root == null) {
            return 0;
        } else {
            sum = sum * 10 + root.val;
            if (root.left == null && root.right == null) {
                return sum;
            } else {
                return preorder(root.left, sum) + preorder(root.right, sum);
            }
        }
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.left = null;
        this.right = null;
        this.val = val;
    }
}

File: tree/leetcode_Sum_Root_to_Leaf_Numbers.py
"""Sum root to leaf numbers
leetcode

Given a binary tree containing digits from 0-9 only, each root-to-leaf path
could represent a number.

An example is the root-to-leaf path 1->2->3 which represents the number 123.

Find the total sum of all root-to-leaf numbers.

For example,

    1
   / \
  2   3
The root-to-leaf path 1->2 represents the number 12.
The root-to-leaf path 1->3 represents the number 13.

Return the sum = 12 + 13 = 25.
"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        total = 0
        queue = [(root, 0)]
        while queue:
            node, parent_sum = queue.pop()

            current_sum = parent_sum * 10 + node.val
            if node.left is None and node.right is None:
                total += current_sum

            if node.left is not None:
                queue.append((node.left, current_sum))
            if node.right is not None:
                queue.append((node.right, current_sum))

        return total


def main():
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)

    assert sol.sumNumbers(root) == 25


if __name__ == '__main__':
    main()
